crates/functional-yew/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::ops::DerefMut;
+use std::rc::Rc;
+use yew::{Component, ComponentLink, Html, Properties};
+
+thread_local! {
+    static CURRENT_HOOK: Rc<RefCell<Option<Rc<RefCell<HookState>>>>> = Rc::new(RefCell::new(None));
+}
+
+struct HookState {
+    counter: usize,
```

Comment:
Then things do ideed start breaking. A possible way around this is to turn hooks into macros that respect the current line number the hook is called from, at a small perfomance cost (which I think is still acceptable).
React has that same restriction, but they also don't have a line number macro ;)

Reply:
clever, I dig that!

Reply:
Thanks, but that would strongly limit custom hooks, since custom hooks would then need to be macros too.
Otherwise if a macro is defined in a separate file but used twice, but always in a different order, there would be no way to know what component line number is calling us. So maybe we could add an extra line insensitive macro later

Reply:
Keeping this in mind for later

Reply:
@ZainlessBrombie I'm thinking it would be pretty nice to relax the ordering restriction of hooks with the line column and filename macros. Since declarative macros are sufficient, I think having all hooks be macros is fine. We could even have a macro which helps create hook macros :)

The case I'm thinking about is inline custom hooks / context hooks used inside the html macro.